src/core/ext/transport/chttp2/server/chttp2_server.cc

```diff
@@ -950,20 +968,647 @@ void Chttp2ServerListener::Orphan() {
+        grpc_transport_op* op = grpc_make_transport_op(nullptr);
+        op->disconnect_with_error = GRPC_ERROR_CREATE(
+            "Did not receive HTTP/2 settings before handshake timeout");
+        self->connection_->transport_->PerformOp(op);
+      },
+      DEBUG_LOCATION);
+}
+
+void NewChttp2ServerListener::NewActiveConnection::HandshakingState::
+    OnReceiveSettings(void* arg, grpc_error_handle /* error */) {
+  HandshakingState* self = static_cast<HandshakingState*>(arg);
+  self->connection_->work_serializer()->Run(
+      [self] {
+        if (self->timer_handle_.has_value()) {
+          self->connection_->event_engine()->Cancel(*self->timer_handle_);
+          self->timer_handle_.reset();
+        }
+        self->Unref();
+      },
+      DEBUG_LOCATION);
+}
+
+void NewChttp2ServerListener::NewActiveConnection::HandshakingState::
+    OnHandshakeDoneLocked(absl::StatusOr<HandshakerArgs*> result) {
+  OrphanablePtr<HandshakingState> handshaking_state_ref;
+  RefCountedPtr<HandshakeManager> handshake_mgr;
+  bool cleanup_connection = false;
+  bool release_connection = false;
+  if (!result.ok() || connection_->shutdown_) {
+    cleanup_connection = true;
+    release_connection = true;
+  } else {
+    // If the handshaking succeeded but there is no endpoint, then the
+    // handshaker may have handed off the connection to some external
+    // code, so we can just clean up here without creating a transport.
+    if ((*result)->endpoint != nullptr) {
+      RefCountedPtr<Transport> transport =
+          grpc_create_chttp2_transport((*result)->args,
+                                       std::move((*result)->endpoint), false)
+              ->Ref();
+      grpc_error_handle channel_init_err =
+          connection_->listener_->listener_wrapper_->server()->SetupTransport(
+              transport.get(), accepting_pollset_, (*result)->args,
+              grpc_chttp2_transport_get_socket_node(transport.get()));
+      if (channel_init_err.ok()) {
+        // Use notify_on_receive_settings callback to enforce the
+        // handshake deadline.
+        connection_->transport_ =
+            DownCast<grpc_chttp2_transport*>(transport.get())->Ref();
+        Ref().release();  // Held by OnReceiveSettings().
+        GRPC_CLOSURE_INIT(&on_receive_settings_, OnReceiveSettings, this,
+                          grpc_schedule_on_exec_ctx);
+        // If the listener has been configured with a config fetcher, we
+        // need to watch on the transport being closed so that we can update
+        // the list of active connections.
+        grpc_closure* on_close = nullptr;
+        if (connection_->listener_->listener_wrapper_->server()
+                ->config_fetcher() != nullptr) {
+          // Refs helds by OnClose()
+          connection_->Ref().release();
+          on_close = &connection_->on_close_;
+        } else {
+          // Remove the connection from the connections_ map immediately since
```

Comment:
This comment needs to be updated, since there is no longer a `connections_` map here.